Add Lex slot type data source and resource

jzbruno opened the pull request:
<!--- See what makes a good Pull Request at : https://github.com/terraform-providers/terraform-provider-aws/blob/master/.github/CONTRIBUTING.md#pull-requests --->

<!--- Please keep this note for the community --->

### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" comments, they generate extra noise for pull request followers and do not help prioritize the request

<!--- Thank you for keeping this note for the community --->

Relates #905 and split from PR #2616

Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md):
<!--
If change is not user facing, just write "NONE" in the release-note block below.
-->

```release-note
FEATURES:
* **New Data Source:** `aws_lex_slot_type`
* **New Resource:** `aws_lex_slot_type`
```

Output from acceptance testing:

```
$ make testacc TESTARGS='-run=TestAcc.*Lex.*'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./... -v -parallel 20 -run=TestAcc.*Lex.* -timeout 120m
?   	github.com/terraform-providers/terraform-provider-aws	[no test files]
=== RUN   TestAccDataSourceAwsLexSlotType
--- PASS: TestAccDataSourceAwsLexSlotType (12.52s)
=== RUN   TestAccLexSlotType_basic
=== PAUSE TestAccLexSlotType_basic
=== RUN   TestAccAwsLexSlotType_CreateVersion
=== PAUSE TestAccAwsLexSlotType_CreateVersion
=== RUN   TestAccAwsLexSlotType_Description
=== PAUSE TestAccAwsLexSlotType_Description
=== RUN   TestAccAwsLexSlotType_EnumerationValue
=== PAUSE TestAccAwsLexSlotType_EnumerationValue
=== RUN   TestAccAwsLexSlotType_Name
=== PAUSE TestAccAwsLexSlotType_Name
=== RUN   TestAccAwsLexSlotType_ValueSelectionStrategy
=== PAUSE TestAccAwsLexSlotType_ValueSelectionStrategy
=== CONT  TestAccLexSlotType_basic
=== CONT  TestAccAwsLexSlotType_Name
=== CONT  TestAccAwsLexSlotType_ValueSelectionStrategy
=== CONT  TestAccAwsLexSlotType_EnumerationValue
=== CONT  TestAccAwsLexSlotType_Description
=== CONT  TestAccAwsLexSlotType_CreateVersion
--- PASS: TestAccLexSlotType_basic (9.37s)
--- PASS: TestAccAwsLexSlotType_CreateVersion (18.77s)
--- PASS: TestAccAwsLexSlotType_Description (19.74s)
--- PASS: TestAccAwsLexSlotType_EnumerationValue (21.03s)
--- PASS: TestAccAwsLexSlotType_Name (22.03s)
--- PASS: TestAccAwsLexSlotType_ValueSelectionStrategy (23.35s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	35.933s
...
```

jzbruno commented on aws/resource_aws_lex_slot_type.go:
```diff
@@ -0,0 +1,184 @@
+func resourceAwsLexSlotTypeRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).lexmodelconn
+
+	version := lexVersionLatest
```
This was here due to import needing a default that doesn't exist in ResourceData at import time. I moved this to the import function instead.

jzbruno commented on aws/resource_aws_lex.go:
```diff
@@ -0,0 +1,69 @@
+package aws
```
@nywilken This file gets pretty large when complete, or are you suggesting the functions go in the resource file they are first used in?

nywilken replied:
Yes - functions that are specific to the resources should go into their respective resource files. It's okay to share functions across the complimenting data_source/resource. If the function needs to cross resources of the same service that is also okay. The idea is centralize the functions where they are first needed.

jzbruno commented on aws/resource_aws_lex.go:
```diff
@@ -0,0 +1,69 @@
+// value pairs and not complex or mixed types. That is why these resources are defined using the
+// schema.TypeList and a max of 1 item instead of the schema.TypeMap.
+
+func expandLexSet(s *schema.Set) (items []map[string]interface{}) {
```
@nywilken This function is used 13 times in the complete code, for intents, slots, enumeration values, messages. Do you still want me to move the code to each of those expand functions?

jzbruno commented:
@nywilken Except for the question above, I have made the requested changes to all 4 PRs.

jzbruno commented:
@nywilken Ok. I have moved all common code into the resource that first uses them, and we will deal with the conflicts as each are merged. Waiting on response about moving expandLexSet into each of the 13 calls to it.

nywilken commented on aws/resource_aws_lex_slot_type.go:
```diff
@@ -0,0 +1,255 @@
+					},
+				},
+			},
+			"name": {
```
We should add a test for this argument separate from the basic test. 

```go
func TestAccAwsLexSlotType_Name(t *testing.T) {                                                                                                   
  resourceName := "aws_lex_slot_type.test"                                                                                                        
  testID := acctest.RandStringFromCharSet(8, acctest.CharSetAlpha)                                                                                
  testSlotTypeID1 := "test_slot_type_" + testID                                                                                                   
  testSlotTypeID2 := "test_slot_type_" + testID                                                                                                   
                                                                                                                                                  
  resource.ParallelTest(t, resource.TestCase{                                                                                                     
    PreCheck:     func() { testAccPreCheck(t) },                                                                                                  
    Providers:    testAccProviders,                                                                                                               
    CheckDestroy: testAccCheckAwsLexSlotTypeDestroy(testSlotTypeID1, "$LATEST"),                                                                  
    Steps: []resource.TestStep{                                                                                                                   
      {                                                                                                                                           
        Config: fmt.Sprintf(testAccAwsLexSlotTypeConfig, testSlotTypeID1),                                                                                 
        Check: resource.ComposeAggregateTestCheckFunc(                                                                                            
          testAccCheckAwsLexSlotTypeExists(testSlotTypeID1, "$LATEST"),                                                                           
          resource.TestCheckResourceAttr(resourceName, "description", "Types of flowers to pick up"),                                             
          resource.TestCheckResourceAttr(resourceName, "enumeration_value.#", "1"),                                                               
          resource.TestCheckResourceAttr(resourceName, "name", testSlotTypeID1),                                                                  
        ),                                                                                                                                        
      },                                                                                                                                          
      {                                                                                                                                           
        ResourceName:      resourceName,                                                                                                          
        ImportState:       true,                                                                                                                  
        ImportStateVerify: true,                                                                                                                  
      },                                                                                                                                          
      {                                                                                                                                           
        Config: fmt.Sprintf(testAccAwsLexSlotTypeConfig, testSlotTypeID2),                                                                                 
        Check: resource.ComposeAggregateTestCheckFunc(                                                                                            
          testAccCheckAwsLexSlotTypeExists(testSlotTypeID2, "$LATEST"),                                                                           
          resource.TestCheckResourceAttr(resourceName, "description", "Types of flowers to pick up"),                                             
          resource.TestCheckResourceAttr(resourceName, "enumeration_value.#", "1"),                                                               
          resource.TestCheckResourceAttr(resourceName, "name", testSlotTypeID2),                                                                  
        ),                                                                                                                                        
      },                                                                                                                                          
      {                                                                                                                                           
        ResourceName:      resourceName,                                                                                                          
        ImportState:       true,                                                                                                                  
        ImportStateVerify: true,                                                                                                                  
      },                                                                                                                                          
    },                                                                                                                                            
  })                                                                                                                                              
} 
```

nywilken commented on aws/resource_aws_lex_slot_type.go:
```diff
@@ -0,0 +1,255 @@
+					validation.StringMatch(regexp.MustCompile(`^((AMAZON\.)_?|[A-Za-z]_?)+`), ""),
+				),
+			},
+			"value_selection_strategy": {
```
We need a test for this argument as well. See my comment below :point_down:

nywilken commented on aws/resource_aws_lex_slot_type_test.go:
```diff
@@ -0,0 +1,146 @@
+    value = "lilies"
+  }
+
+  name                     = "test_slot_type_%s"
```
Since you are passing in TestId above which contains `test_slot_type_` you can simplify this to 
```suggestion
  name                     = %q
```

nywilken commented on aws/resource_aws_lex_slot_type_test.go:
```diff
@@ -0,0 +1,146 @@
+		CheckDestroy: testAccCheckAwsLexSlotTypeDestroy(testSlotTypeID, "$LATEST"),
+		Steps: []resource.TestStep{
+			{
+				Config: fmt.Sprintf(testAccAwsLexSlotTypeConfig, testID),
```
I left a suggestion below in using `%q` for the config. If you accept that suggestion then you will need to change `testId` to the expected name `testSlotTypeID`

```suggestion
				Config: fmt.Sprintf(testAccAwsLexSlotTypeConfig, testSlotTypeID),
```

nywilken commented on aws/resource_aws_lex_slot_type_test.go:
```diff
@@ -0,0 +1,146 @@
+}
+`
+
+const testAccAwsLexSlotTypeUpdateConfig = `
```
This is a nit, but usually when creating an update test config we specify what is being updated for example this would be `testAccAwsLexSlotTypeUpdateEnumerationValuesConfig`

jzbruno commented:
> @jzbruno thanks for you patience on the review, I had a little more time to dive into the API and get a sense of what is expected for this resource. I left a few implementation suggestions to address your questions, and to improve the Acceptance testing.
> 
> There is a question around the need for the version attribute within the schema which from my testing always appears to be set to `$LATEST`. Please take a look and let me know if you have any questions. I will start to review the other PRs now that I have a sense of what is happening with this service and this one is just about to be done.
> 
> Thanks again for your patience and support in pushing this service forward.

@nywilken So, version doesn't always have to be $LATEST, you could for example manage multiple versions of bots, intents, and slot types. Let's say you create v1 and then roll out v2 but v1 is still active you would still want to manage both. Another example would be having a dev and prod version.

https://docs.aws.amazon.com/lex/latest/dg/versioning-aliases.html

What do you think?

jzbruno commented:
@nywilken 
Will wait on version until we discuss further. I updated the PR with the other suggestions.

How far do we want to go with breaking apart the tests for the other resources? For the Intent resource there 56 different update combinations due to the multiple levels of nested resources.

jzbruno commented:
@nywilken After reviewing the version issue further I think you are correct. The SlotType resource should always track the latest version so we can remove that as an input. I will add the create_version attribute to allow enable/disable of the version snapshotting for each change. And also add a resource for SlotTypeVersion to allow management of those version snapshots.

jzbruno commented:
Ok. This is ready for review again. Should handle version correctly. 

Only issue is dealing with attributes that are used for create / update but don't get returned from the API like create_version. They cause issues with importverifystate testing since the config isn't loaded at that stage and read can't get it from the API. Makes it so we have to disable the import verify on that particular test case. Any suggestions there?

nywilken commented on aws/resource_aws_lex_slot_type_test.go:
```diff
@@ -0,0 +1,348 @@
+	"github.com/hashicorp/terraform/terraform"
+)
+
+func TestAccLexSlotType_basic(t *testing.T) {
```
```suggestion
func TestAccAwsLexSlotType_basic(t *testing.T) {
```

nywilken commented on aws/resource_aws_lex_slot_type_test.go:
```diff
@@ -0,0 +1,348 @@
+			{
+				ResourceName: resourceName,
+				ImportState:  true,
+				// Cannot verify import for create_version other than true because we don't have
```
The field you are looking for here is `ImportStateVerifyIgnore` which allows you to specify which resource arguments can be ignored in the verification step since they are not part of the API response. 

```suggestion
ImportStateVerifyIgnore: []string{"create_version"}
```

nywilken commented on website/docs/r/lex_slot_type.html.markdown:
```diff
@@ -0,0 +1,125 @@
+	learning model about the values that it resolves for a slot. Attributes are documented under
+  [enumeration_value](#enumeration_value-1).
+
+* `name` _(**Required**)_:
```
These don't need to be bold and in italics. The Required/Optional keywords do not get styled in anyway other than being wrapped in parentheses. The description for the argument should be inline.
 
```suggestion
* `name` (Required): The name of the slot type. The name is not case sensitive.
```

nywilken commented on website/docs/r/lex_slot_type.html.markdown:
```diff
@@ -0,0 +1,125 @@
+
+	The name of the slot type. The name is not case sensitive.
+
+* `value_selection_strategy` _(Optional)_:
```
Same comments as above. No styles necessary, and the description should be inline.

jzbruno commented:
@nywilken Ok. Updated for your comments. This should be good for another review. Thanks. And I will start work on the other PRs now.

jzbruno commented:
An update for anyone following. The updates to intent are taking longer than expected. I will also be out of town through Sunday but will start working on it again next week. Sorry for the delay.

scott-robinson-duke commented:
This doesn't look to be part of v2.19.0 at the moment.  Maybe v2.20.0? I assume the intent code will handle lambda initialization and/or fulfillment?  Thanks.

nywilken commented:
@jzbruno I had to take some time off so I've not had a chance to follow up since your last commit. I'll be starting fresh on Monday with the reviews. Thanks again for pushing the Lex resources forward, and for sticking with me.

scott-robinson-duke commented:
This doesn't look to be part of v2.20.0 at the moment. Maybe v2.21.0? I assume the intent code will handle lambda initialization and/or fulfillment? Thanks.

scott-robinson-duke commented:
Is this likely to be a part of v2.21.0? Really, really need this update!  Thanks.

jzbruno commented:
@scott-robinson-duke Sorry, I have not had time to continue working on this. I think I am still waiting on some review as well.

As for Lambda. The intent can link to a Lambda, that is how the AWS SDK is setup. Those Lambda's would need to be setup through the Lambda Terraform resource I believe.

scott-robinson-duke commented:
OK.  Trying to plan.  Is it possible to get an ETA of some sort?  Thanks.

nywilken commented:
@jzbruno, and folks following along, my apologies for the longer than normal delays on pushing forward this feature. I haven't had the chance to devote the needed time get the other pull-requests into a mergable state 😳 I appreciate your patience as we continue to work through this resource. 

While I can't say that it will get into the next upcoming release(s), I can say that the resource is still a priority for us. 

To not give false expectations around the release of this resource I am going to remove the milestone. But let's coordinate on when would be ideal to continue pushing forward. If you find that your availability has also slowed please let me know if we should consider having another community member or the maintainers take over the pull-request(s). 

Cheers

jzbruno commented:
@nywilken I can continue to work on this. I believe this is ready for further review.

scott-robinson-duke commented:
Any updates?  Thanks in advance!

leonfs commented:
@nywilken Any updates?

jzbruno commented:
> Hi @jzbruno, our apologies for the delay on reviewing this and the other Lex PRs you've contributed and thank you for your patience. You've done a lot of good work on the PRs!
> 
> Are you still interested in working on these PRs after the long delay? If not, we can finish them. If you're still interested, let me know by 24 June

Awesome! Yes, I will work on the review comments this weekend.

gdavison commented:
I'm also going to reverse the suggestion about removing the shared structures and functions file. Since it's used in so many places, I think it makes sense to move them to a common file. Maybe rename it to `lex.go`, since the pattern `resource_aws_<something>` suggests that it's a Terraform resource

gdavison commented:
Hi @jzbruno, have you had a chance to look at this?

jzbruno commented:
Working through resolving conflicts and getting it into a buildable state now. Will address your comment after that. Hopefully have it up tonight.

jzbruno commented on aws/resource_aws_lex_slot_type.go:
```diff
@@ -0,0 +1,249 @@
+	}
+	return enums
+}
+
```
Would you prefer one per resource type or a global lextVersionLatest? This is used in every lex resource if I remember.

jzbruno replied:
Went ahead and just used LexSlotTypeVersionLatest for now.

jzbruno commented on aws/resource_aws_lex_slot_type.go:
```diff
@@ -0,0 +1,249 @@
+	"github.com/hashicorp/terraform/helper/validation"
+)
+
+// Many of the Lex resources require complex nested objects. Terraform maps only support simple key
+// value pairs and not complex or mixed types. That is why these resources are defined using the
+// schema.TypeList and a max of 1 item instead of the schema.TypeMap.
```
Yeah been a while think this was leftover  after first set of review changes.

jzbruno commented on aws/resource_aws_lex_slot_type.go:
```diff
@@ -0,0 +1,249 @@
+	d.Set("description", resp.Description)
+	d.Set("name", resp.Name)
+	d.Set("value_selection_strategy", resp.ValueSelectionStrategy)
+	d.Set("version", "$LATEST")
```
So, I believe the problem is that if we store the actual version in state and the user has $LATEST in their config, it will always show a diff on apply. How should we handle this?

gdavison replied:
Since the `version` value is read-only, this shouldn't be a problem

jzbruno commented:
Ok. Calling it for the night. Hacked through a lot of it. Left a couple questions. Will pick it back up tomorrow. :)

jzbruno commented on aws/resource_aws_lex_slot_type.go:
```diff
@@ -0,0 +1,249 @@
+			},
+			"enumeration_value": {
+				Type:     schema.TypeSet,
+				Optional: true,
```
Looks like this was changed but oddly, we already require at least 1. Test cases just didn't adhere to that.

gdavison replied:
```suggestion
				Required: true,
```

jzbruno commented:
Addressed all comments except 2. Do you have a suggestion for the dependency check failing?

gdavison commented:
The dependency check is flagging the updates to `go.mod`, `go.sum`, and `vendor/`. Since this PR has been around for a while, you probably had to make a number of updates. The easiest way to work around this would be to either rebase your branch on the current main branch or to merge the current main branch into your PR. Github will see the changes either way. You may also need to run `go mod tidy` once or twice to remove changes to `go.mod`

jzbruno commented:
Ok, thanks. I will try to resolve that and finish the other 2 comments tonight. Sorry for the delay on this.

jzbruno commented:
@gdavison I believe I have addressed all of your comments. Latest acc test run shows

gdavison commented on aws/resource_aws_lex_slot_type_test.go:
```diff
@@ -0,0 +1,402 @@
+	})
+}
+
+func testAccCheckAwsLexSlotTypeExistsWithVersion(slotTypeName, slotTypeVersion string, output *lexmodelbuildingservice.GetSlotTypeOutput) resource.TestCheckFunc {
+	return func(s *terraform.State) error {
+		var err error
+		conn := testAccProvider.Meta().(*AWSClient).lexmodelconn
+
+		output, err = conn.GetSlotType(&lexmodelbuildingservice.GetSlotTypeInput{
+			Name:    aws.String(slotTypeName),
+			Version: aws.String(slotTypeVersion),
+		})
+		if isAWSErr(err, lexmodelbuildingservice.ErrCodeNotFoundException, "") {
+			return fmt.Errorf("error slot type %s not found, %s", slotTypeName, err)
+		}
+		if err != nil {
+			return fmt.Errorf("error getting slot type %s: %s", slotTypeName, err)
+		}
+
+		return nil
+	}
+}
```
The pattern we use for `Exists` functions is to pass the resource address, e.g. `aws_lex_slot_type.test` and then look up relevant information from the Terraform state to call the AWS API.

Something like the following. This also specifies which version isn't found in the error messages for completeness
```go
func testAccCheckAwsLexSlotTypeExistsWithVersion(n, slotTypeVersion string, res *lexmodelbuildingservice.GetSlotTypeOutput) resource.TestCheckFunc {
	return func(s *terraform.State) error {
		rs, ok := s.RootModule().Resources[n]
		if !ok {
			return fmt.Errorf("Not found: %s", n)
		}

		if rs.Primary.ID == "" {
			return fmt.Errorf("No Lex slot type ID is set")
		}

		conn := testAccProvider.Meta().(*AWSClient).lexmodelconn

		output, err := conn.GetSlotType(&lexmodelbuildingservice.GetSlotTypeInput{
			Name:    aws.String(rs.Primary.ID),
			Version: aws.String(slotTypeVersion),
		})
		if isAWSErr(err, lexmodelbuildingservice.ErrCodeNotFoundException, "") {
			return fmt.Errorf("error slot type %q version %s not found", rs.Primary.ID, slotTypeVersion)
		}
		if err != nil {
			return fmt.Errorf("error getting slot type %q version %s: %w", rs.Primary.ID, slotTypeVersion, err)
		}

		*res = *output

		return nil
	}
}
```

gdavison commented on aws/resource_aws_lex_slot_type_test.go:
```diff
@@ -0,0 +1,402 @@
+			return nil
+		}
+		if err != nil {
+			return fmt.Errorf("error getting slot type %s: %s", slotTypeName, err)
+		}
+
+		return fmt.Errorf("error slot type exists %s", slotTypeName)
```
Error messages should include the version

gdavison commented on aws/resource_aws_lex_slot_type_test.go:
```diff
@@ -0,0 +1,402 @@
+    ]
+    value = "lilies"
+  }
+  create_version = true
```
This configuration should not set `create_version`. In general, the `_basic` configuration should use as many default values as possible

gdavison commented on website/docs/r/lex_slot_type.html.markdown:
```diff
@@ -0,0 +1,97 @@
+* `checksum` - Checksum identifying the version of the slot type that was created. The checksum is 
+not included as an argument because the resource will add it automatically when updating the slot type.
+* `created_date` - The date when the slot type version was created.
+* `last_updated_date` - The date when the $LATEST version of this slot type was updated.
```
We can style the constant value
```suggestion
* `last_updated_date` - The date when the `$LATEST` version of this slot type was updated.
```

gdavison commented on website/docs/r/lex_slot_type.html.markdown:
```diff
@@ -0,0 +1,97 @@
+
+The following arguments are supported:
+
+* `create_version` - (Optional) 
+Determines if a new slot type version is created when the initial resource is created and on each 
+update. Defaults to true.
+* `description` - (Optional) A description of the slot type.
+* `enumeration_value` - (Required) A list of EnumerationValue objects that defines the values that 
+the slot type can take. Each value can have a list of synonyms, which are additional values that help 
+train the machine learning model about the values that it resolves for a slot. Attributes are 
+documented under [enumeration_value](#enumeration_value-1).
+* `name` - (Required) The name of the slot type. The name is not case sensitive.
+* `value_selection_strategy` - (Optional) Determines the slot resolution strategy that Amazon Lex 
+uses to return slot type values. `ORIGINAL_VALUE` returns the value entered by the user if the user 
+value is similar to the slot value. `TOP_RESOLUTION` returns the first value in the resolution list 
+if there is a resolution list for the slot, otherwise null is returned.
```
For readability, it might be good to move the `Required` fields to the top

gdavison commented on website/docs/d/lex_slot_type.html.markdown:
```diff
@@ -0,0 +1,32 @@
+The following arguments are supported:
+
+* `name` - (Required) The name of the slot type. The name is case sensitive.
+* `version` - (Required) The version of the slot type.
```
This should be updated, since the `version` is optional, with the default of `$LATEST`

gdavison commented on website/docs/d/lex_slot_type.html.markdown:
```diff
@@ -0,0 +1,32 @@
+
+## Attributes Reference
+
+All attributes are exported. See the [aws_lex_slot_type](/docs/providers/aws/r/lex_slot_type.html) 
+resource for the full list.
```
We should list the output values here, rather than linking to the resource documentation

jzbruno commented:
Working through these. Will finish up tonight.

jzbruno commented on aws/resource_aws_lex_slot_type_test.go:
```diff
@@ -0,0 +1,402 @@
+	resource.ParallelTest(t, resource.TestCase{
+		PreCheck:     func() { testAccPreCheck(t) },
+		Providers:    testAccProviders,
+		CheckDestroy: testAccCheckAwsLexSlotTypeNotExists(testSlotTypeID, LexSlotTypeVersionLatest),
```
Are we sure it's possible for this state to occur? I thought if you delete the slot type all versions are deleted, so you should just be able to check for it's existence. And we don't provide a way to delete versions with this resource.

gdavison replied:
Sorry, it's been a while since I looked at this. I believe I ran into problems if there were multiple published versions. In that case, the latest version could be deleted but older versions would not be.

jzbruno commented on aws/resource_aws_lex_slot_type.go:
```diff
@@ -0,0 +1,301 @@
+		})
+
+		if isAWSErr(err, lexmodelbuildingservice.ErrCodeConflictException, "") {
+			return resource.RetryableError(fmt.Errorf("%q: slot type still deleting", d.Id()))
```
Are you seeing an error deleting? I have never seen an issue here. See my comments below about the versions not being able to exist without the slot type itself.

gdavison replied:
Yes, because there is a slight delay in deletion, especially when there are multiple versions, I was occasionally hitting errors. The two functions described above are a standard pattern that we've adopted for handling this.

jzbruno commented:
Addressed most comments but had a couple questions. Latest acceptance test results.

```
~/src/github/terraform-providers/terraform-provider-aws $ make testacc TEST=./aws TESTARGS='-run=TestAcc.*AwsLexSlotType.*'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAcc.*AwsLexSlotType.* -timeout 120m
=== RUN   TestAccDataSourceAwsLexSlotType_basic
=== PAUSE TestAccDataSourceAwsLexSlotType_basic
=== RUN   TestAccDataSourceAwsLexSlotType_withVersion
=== PAUSE TestAccDataSourceAwsLexSlotType_withVersion
=== RUN   TestAccAwsLexSlotType_basic
=== PAUSE TestAccAwsLexSlotType_basic
=== RUN   TestAccAwsLexSlotType_createVersion
=== PAUSE TestAccAwsLexSlotType_createVersion
=== RUN   TestAccAwsLexSlotType_description
=== PAUSE TestAccAwsLexSlotType_description
=== RUN   TestAccAwsLexSlotType_enumerationValues
=== PAUSE TestAccAwsLexSlotType_enumerationValues
=== RUN   TestAccAwsLexSlotType_name
=== PAUSE TestAccAwsLexSlotType_name
=== RUN   TestAccAwsLexSlotType_valueSelectionStrategy
=== PAUSE TestAccAwsLexSlotType_valueSelectionStrategy
=== RUN   TestAccAwsLexSlotType_disappears
=== PAUSE TestAccAwsLexSlotType_disappears
=== CONT  TestAccDataSourceAwsLexSlotType_basic
=== CONT  TestAccAwsLexSlotType_enumerationValues
=== CONT  TestAccAwsLexSlotType_createVersion
=== CONT  TestAccAwsLexSlotType_name
=== CONT  TestAccAwsLexSlotType_basic
=== CONT  TestAccAwsLexSlotType_valueSelectionStrategy
=== CONT  TestAccAwsLexSlotType_disappears
=== CONT  TestAccAwsLexSlotType_description
=== CONT  TestAccDataSourceAwsLexSlotType_withVersion
--- PASS: TestAccAwsLexSlotType_disappears (14.57s)
--- PASS: TestAccAwsLexSlotType_basic (19.99s)
--- PASS: TestAccDataSourceAwsLexSlotType_basic (20.33s)
--- PASS: TestAccDataSourceAwsLexSlotType_withVersion (26.15s)
--- PASS: TestAccAwsLexSlotType_enumerationValues (34.64s)
--- PASS: TestAccAwsLexSlotType_createVersion (35.31s)
--- PASS: TestAccAwsLexSlotType_description (35.99s)
--- PASS: TestAccAwsLexSlotType_valueSelectionStrategy (36.90s)
--- PASS: TestAccAwsLexSlotType_name (42.22s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	43.845s
```

jzbruno commented:
@gdavison Thanks I will take a look this weekend.

jzbruno commented:
@gdavison I think I finished all your suggestions.

leonfs commented:
This is great news. @jzbruno Great work on getting the Lex Slot merged and deployed. Hopefully, the other 3 PRs will progress smoothly too.

New Resource: aws_lex_intent #8917
New Resource: aws_lex_bot #8918
New Resource: aws_lex_bot_alias #8919
